[masonry] Expose grid items positions and sizes to Web Inspector

stwrt opened the pull request:
#### 3816458b86b6fc9a5c62f8dc1d3e96b5498ec554
<pre>
[masonry] Expose grid items positions and sizes to Web Inspector
<a href="https://bugs.webkit.org/show_bug.cgi?id=267736">https://bugs.webkit.org/show_bug.cgi?id=267736</a>
<a href="https://rdar.apple.com/problem/121219933">rdar://problem/121219933</a>

Reviewed by Sammy Gill.

Masonry differs from Grid in only having one axis. This does not allow simple drawing
straight lines. The Web Inspector can instead draw boxes around each item, which this patch
is exposing in a simpler format.

This patch is not exclusive to masonry and can be used for all returning all grid items
layout rects.

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::gridItemsLayoutRects):
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: <a href="https://commits.webkit.org/273232@main">https://commits.webkit.org/273232@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7419dad1cb11f0d9a00b4954b9d8ad7051acf25

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13498 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36683 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37359 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31312 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35817 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15900 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10592 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30272 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35201 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11420 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30892 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9983 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10055 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30969 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38631 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31479 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31281 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36112 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10177 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8070 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34075 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12001 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7988 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10725 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11056 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1022,6 +1022,16 @@ LayoutUnit RenderGrid::masonryContentSize() const
     return m_masonryLayout.gridContentSize();
 }
 
+Vector<LayoutRect> RenderGrid::masonryItems()
```
I think we should rename this function since in its current form it's kinda misleading. We are not actually returning the actual masonry items to the caller but their `LayoutRects` so we should make that clear

stwrt replied:
Thanks. Fixed.

rcaliman-apple reviewed (approved):
Very neat!

These metrics will be helpful in https://github.com/WebKit/WebKit/blob/8c2f912a0edfdb7d2833a620529a49a589711a34/Source/WebCore/inspector/InspectorOverlay.cpp#L1494

to follow-up on [Bug 265381](https://bugs.webkit.org/show_bug.cgi?id=265381) by improving the drawing for items in a masonry layout. 

Thank you!

sammygill commented on Source/WebCore/rendering/RenderGrid.h:
```diff
@@ -128,6 +128,7 @@ class RenderGrid final : public RenderBlock {
     LayoutUnit gridGap(GridTrackSizingDirection, std::optional<LayoutUnit> availableSize) const;
 
     LayoutUnit masonryContentSize() const;
+    Vector<LayoutRect> gridItemsLayoutRects();
```
Let's make this a const function

stwrt replied:
Discuss with Sammy.

This will require changing the orderiterator to support const better.

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1022,6 +1022,16 @@ LayoutUnit RenderGrid::masonryContentSize() const
 
+Vector<LayoutRect> RenderGrid::gridItemsLayoutRects()
+{
+    Vector<LayoutRect> items;
```
Should we assert here to make sure the caller does not call this function on a grid that needs to go through layout? Otherwise they would be getting stale geometry for the renderers

stwrt replied:
Discussed with Sammy about the assert.

This is somewhat more of a discussion who owns the responsibility of ensuring that the data is not stale.

Will handle this in a future patch.

stwrt committed:
[masonry] Expose grid items positions and sizes to Web Inspector
https://bugs.webkit.org/show_bug.cgi?id=267736
rdar://problem/121219933

Reviewed by Sammy Gill.

Masonry differs from Grid in only having one axis. This does not allow simple drawing
straight lines. The Web Inspector can instead draw boxes around each item, which this patch
is exposing in a simpler format.

This patch is not exclusive to masonry and can be used for all returning all grid items
layout rects.

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::gridItemsLayoutRects):
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: https://commits.webkit.org/273232@main
